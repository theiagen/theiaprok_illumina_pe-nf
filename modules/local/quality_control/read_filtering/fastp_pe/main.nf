process FASTP_PE {
    tag "$meta.id"
    label "process_low"

    container "us-docker.pkg.dev/general-theiagen/staphb/fastp:0.23.2"

    input:
    tuple val(meta), path(read1), path(read2)
    val fastp_window_size
    val fastp_quality_trim_score
    val fastp_min_length
    val fastp_args

    output:
    tuple val(meta), path("*P.fastq.gz"), emit: trimmed_reads
    tuple val(meta), path("*U.fastq.gz"), emit:  unpaired_trimmed_reads
    tuple val(meta), path("*_fastp.html"), emit: fastp_stats
    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def fastp_args = fastp_args ?: "--detect_adapter_for_pe -g -5 20 -3 20"
    def prefix = task.ext.prefix ?: "${meta.id}"
    def window_size = fastp_window_size ?: 20
    def quality_score = fastp_quality_trim_score ?: 30
    def min_length = fastp_min_length ?: 50
    
    """
    date | tee DATE
    
    fastp \\
        --in1 ${read1} \\
        --in2 ${read2} \\
        --out1 ${prefix}_1P.fastq.gz \\
        --out2 ${prefix}_2P.fastq.gz \\
        --unpaired1 ${prefix}_1U.fastq.gz \\
        --unpaired2 ${prefix}_2U.fastq.gz \\
        --cut_right \\
        --cut_right_window_size ${window_size} \\
        --cut_right_mean_quality ${quality_score} \\
        --length_required ${min_length} \\
        --thread ${task.cpus} \\
        ${fastp_args} \\
        --html ${prefix}_fastp.html \\
        --json ${prefix}_fastp.json

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        fastp: \$(fastp --version 2>&1 | sed 's/fastp //')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}_1P.fastq.gz
    touch ${prefix}_2P.fastq.gz
    touch ${prefix}_1U.fastq.gz
    touch ${prefix}_2U.fastq.gz
    touch ${prefix}_fastp.html

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        fastp: 0.23.2
    END_VERSIONS
    """
}