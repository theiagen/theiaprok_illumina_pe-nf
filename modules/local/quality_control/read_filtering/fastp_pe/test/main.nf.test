nextflow_process {
    name "Test Process FASTP_PE"
    script "../main.nf"
    process "FASTP_PE"
    
    test("fastp_pe test") {
        when {
            process {
                """
                input[0] = [
                    [ id:'test_sample' ],
                    file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz", checkIfExists: true),
                    file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_2.fastq.gz", checkIfExists: true)
                ]
                input[1] = 20  // fastp_window_size
                input[2] = 30  // fastp_quality_trim_score
                input[3] = 50  // fastp_min_length
                input[4] = "--detect_adapter_for_pe -g -5 20 -3 20"  // fastp_args
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                {  assert snapshot(
                    process.out.trimmed_reads,
                    process.out.unpaired_trimmed_reads,
                    process.out.fastp_stats,
                    process.out.versions
                    ).match() }
            )
        }
    }
    
    test("fastp_pe - stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = [
                    [ id:'test_sample' ],
                    file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz", checkIfExists: true),
                    file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_2.fastq.gz", checkIfExists: true)
                ]
                input[1] = 20  // fastp_window_size
                input[2] = 30  // fastp_quality_trim_score
                input[3] = 50  // fastp_min_length
                input[4] = "--detect_adapter_for_pe -g -5 20 -3 20"  // fastp_args
                """
            }
        }
        then {
            assertAll(
                { assert process.success }
            )
        }
    }
}