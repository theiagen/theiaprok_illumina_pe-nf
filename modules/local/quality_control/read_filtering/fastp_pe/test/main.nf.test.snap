{
    "fastp_pe test": {
        "content": [
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample_1P.fastq.gz:md5,2c631cf69d30772b5d83fa4050628943"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample_2P.fastq.gz:md5,4e4c1b198549ca88e3c3de4ef6670a93"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample_1U.fastq.gz:md5,af069f7905bba8476f3966a58e3196db"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample_2U.fastq.gz:md5,9d07c0f8872043e4e4a8b7e87f2b4b59"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample_fastp.html:md5,f199171e4e91d0f750ec24e067e0bce6"
                ]
            ],
            [
                "versions.yml:md5,eb4e79b333148a2aacb75085ec45837e"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-06-17T19:12:12.789758"
    }
}